### PR TITLE
fix: fixes transactionID isolation for GraphQL

### LINKED ISSUE #4091: HTTP Headers not available in all GraphQL Resolvers
### Link to reproduction

https://github.com/diesieben07/payload/tree/repro/graphql-request-headers

### Describe the Bug

https://github.com/payloadcms/payload/pull/3985 now clones the request before calling _most_ resolvers (it is missing `collections/findByID`, which is another issue). This is done via `{ ...req }` which loses some properties (including `headers`).

### To Reproduce

Attempt to use `PayloadRequest#headers` within a GraphQL request. Note how it is undefined.

### Workarounds

This can be worked around by the following ugly hack:
```ts
function getHeaders(req: PayloadRequest): IncomingHttpHeaders | null {
  if (req.headers != null) {
    return req.headers;
  } else {
    const headersSymbol = Object.getOwnPropertySymbols(req).find(s => s.description === 'kHeaders');
    if (headersSymbol != null) {
      return (req as never)[headersSymbol] ?? null;
    } else {
      return null;
    }
  }
}
```

### Payload Version

2.1.0

### Adapters and Plugins

n/a

### PR DESCRIPTION
**Pull request by diesieben07**
## Description

Fixes #4091 by using a JavaScript Proxy to isolate the `transactionID` between individual GraphQL operations instead of copying the request.
I have also fixed the missing cloning of the request for the `collections/findByID` resolver.

- [X] I have read and understand the [CONTRIBUTING.md](https://github.com/payloadcms/payload/blob/main/CONTRIBUTING.md) document in this repository.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [X] I have added tests that prove my fix is effective or that my feature works
- [X] Existing test suite passes locally with my changes

### CODE REVIEW

File: packages/payload/src/utilities/isolateTransactionID.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import type { PayloadRequest } from '../express/types'
+
+/**
+ * Creates a proxy for the given request that has its own TransactionID
+ */
+export default function isolateTransactionID(req: PayloadRequest): PayloadRequest {
```
Comment (DanRibbens) [inline packages/payload/src/utilities/isolateTransactionID.ts:6]: I was looking for a way to do this and couldn't find it. Great work!

**Review by DanRibbens — CHANGES_REQUESTED** [review]
If you do a search for `req: { ...context.req } as PayloadRequest` you'll find a few more places where I did this which your PR should also replace.
Looking at: `src/auth/graphql/resolvers`

**Comment by diesieben07** [comment]
@DanRibbens Right, of course :man_facepalming: I've now gone through `registerSchema` and checked every resolver.

Should we expose `isolateTransactionID` for people to use in custom GraphQL resolvers or maybe we should handle it transparently by wrapping any custom resolver with it? The latter sounds more convenient for users and makes it impossible to forget.

**Comment by DanRibbens** [comment]
> Should we expose `isolateTransactionID` for people to use in custom GraphQL resolvers or maybe we should handle it transparently by wrapping any custom resolver with it? The latter sounds more convenient for users and makes it impossible to forget.

We never want people to have to learn these kinds of ugly details if it can be avoided. I haven't reviewed this yet but it sounds like we need to handle it.

Thanks!

**Comment by diesieben07** [comment]
I have now added a wrapper to any (top level) resolver in the custom mutations and queries. I've added a test that makes sure they are using separate transaction IDs.

**Review by DanRibbens — APPROVED** [review]

**Pull request merged**